types/src/sample.rs

```diff
@@ -2,45 +2,153 @@ use std::mem::size_of;
 
-use crate::axis::{AxisId, AxisType};
-use crate::DataAvailabilityHeader;
-use crate::{Error, Result};
+use crate::extended_data_square::{AxisType, ExtendedDataSquare};
+use crate::nmt::{NamespaceProof, NamespacedHash, NamespacedHashExt, NamespacedSha2Hasher, Nmt};
+use crate::row::RowId;
+use crate::{Error, Result, Share};
 
 const SAMPLE_ID_SIZE: usize = SampleId::size();
 pub const SAMPLE_ID_MULTIHASH_CODE: u64 = 0x7801;
 pub const SAMPLE_ID_CODEC: u64 = 0x7800;
 
-/// Represents particular sample along the axis on specific Data Square
+/// Represents a location of a sample on the EDS
 #[derive(Debug, PartialEq, Clone, Copy)]
 pub struct SampleId {
-    pub axis: AxisId,
+    pub row: RowId,
     pub index: u16,
 }
 
-impl SampleId {
-    /// Create new SampleId. Index references sample number from the entire Data Square (is
-    /// converted to row/col coordinates internally). Same location can be sampled row or
-    /// column-wise, axis_type is used to distinguish that. Axis root hash is calculated from the
-    /// DataAvailabilityHeader
-    #[allow(dead_code)] // unused for now
+/// Represents Sample, with proof of its inclusion and location on EDS
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(try_from = "RawSample", into = "RawSample")]
+pub struct Sample {
+    pub sample_id: SampleId,
+
+    pub sample_proof_type: AxisType,
+    pub share: Share,
+    pub proof: NamespaceProof,
+}
+
+impl Sample {
+    /// Create a new Sample. Index references sample number from the entire Data Square
     pub fn new(
         axis_type: AxisType,
         index: usize,
-        dah: &DataAvailabilityHeader,
+        eds: &ExtendedDataSquare,
         block_height: u64,
     ) -> Result<Self> {
-        let square_len = dah.square_len();
+        let square_len = eds.square_len();
 
         let (axis_index, sample_index) = match axis_type {
             AxisType::Row => (index / square_len, index % square_len),
             AxisType::Col => (index % square_len, index / square_len),
         };
 
+        let shares = eds.axis(axis_type, axis_index)?;
+
+        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
+
+        // XXX: are erasure coded shares correctly prefixed with parity namespace?
+        for s in &shares {
+            tree.push_leaf(s.data(), *s.namespace())
+                .map_err(Error::Nmt)?;
+        }
+
+        let proof = NmtNamespaceProof::PresenceProof {
+            proof: tree.build_range_proof(sample_index..sample_index + 1),
+            ignore_max_ns: true,
+        };
+
+        let sample_id = SampleId::new(index, square_len, block_height)?;
+
+        Ok(Sample {
+            sample_id,
+            sample_proof_type: axis_type,
+            share: shares[sample_index].clone(),
+            proof: proof.into(),
+        })
+    }
+
+    /// Validate sample with root hash from ExtendedHeader
+    pub fn validate(&self, hash: Hash) -> Result<()> {
```

Comment:
as this has all the required information to navigate the correct hash in dah, I think requiring a dah here would be more user-friendly

Reply:
Switched to using entire DAH here and in Row 👍